### PR TITLE
retry block processing

### PR DESCRIPTION
**Pull request by macfarla**
Signed-off-by: Justin Florentine <justin+github@florentine.us>

* if MerkleTrieException, don't mark block as INVALID - and mark for retry
* do not use yield.isPresent to indicate whether block processing was successful

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinatorTest.java [hunk 1]
```diff
@@ -200,6 +222,85 @@ public void coinbaseShouldMatchSuggestedFeeRecipient() {
...
+
+    verify(willThrow, never()).addBadBlock(any(), any());
+    verify(mergeContext, times(txPerBlock + 1))
+        .putPayloadById(eq(payloadId), block.capture()); // +1 for the empty
+    assertThat(block.getValue().getBody().getTransactions().size()).isEqualTo(txPerBlock);
+    // this only verifies that adding the bad block didn't happen through the mergeCoordinator, it
+    // still may
+    // be called directly.
+    verify(badBlockManager, never()).addBadBlock(any(), any());
+    verify(willThrow, never()).addBadBlock(any(), any());
```
Comment (fab-10) [inline consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinatorTest.java:301]: this seems a duplication of line 293

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java [hunk 2]
```diff
@@ -185,14 +185,10 @@ private void handleAndLogImportFailure(
...
       LOG.debug("with stack", result.causedBy().get());
-      if (!result.isInternalError()) {
-        badBlockManager.addBadBlock(invalidBlock);
-      }
-
     } else {
       LOG.info("{}. Block {}", result.errorMessage, invalidBlock.toLogString());
-      badBlockManager.addBadBlock(invalidBlock);
     }
+    badBlockManager.addBadBlock(invalidBlock, result.causedBy());
```
Comment (fab-10) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java:191]: before an internal error did not cause the block to be added to the bad list, now it is always added, is this wanted?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/BadBlockManager.java [hunk 3]
```diff
@@ -78,4 +83,12 @@ public void addLatestValidHash(final Hash blockHash, final Hash latestValidHash)
   public Optional<Hash> getLatestValidHash(final Hash blockHash) {
     return Optional.ofNullable(latestValidHashes.getIfPresent(blockHash));
   }
+
+  public boolean isInternalError(final Throwable causedBy) {
```
Comment (fab-10) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/BadBlockManager.java:87]: private?

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadTest.java [hunk 4]
```diff
@@ -237,10 +239,54 @@ public void shouldNotReturnInvalidOnInternalException() {
...
+        .thenReturn(true);
+    when(mergeCoordinator.rememberBlock(any()))
+        .thenReturn(
+            new BlockProcessingResult(Optional.empty(), new MerkleTrieException("missing leaf")));
+
+    var resp = resp(mockPayload(mockHeader, Collections.emptyList()));
+
+    verify(engineCallListener, times(1)).executionEngineCalled();
+    verify(mergeCoordinator, times(0)).addBadBlock(any(), any());
+    // verify mainnetBlockValidator does not add to bad block manager
```
[the remark this answers is not in the document]
Comment (jflo) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadTest.java:264]: Yes, but it was handled elsewhere. removed.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadTest.java [hunk 5]
```diff
@@ -237,10 +239,54 @@ public void shouldNotReturnInvalidOnInternalException() {
...
+    when(mergeCoordinator.rememberBlock(any())).thenThrow(new MerkleTrieException("missing leaf"));
+
+    var resp = resp(mockPayload(mockHeader, Collections.emptyList()));
+
+    verify(engineCallListener, times(1)).executionEngineCalled();
+    verify(mergeCoordinator, never()).addBadBlock(any(), any());
+    // verify mainnetBlockValidator does not add to bad block manager
+
+    fromErrorResp(resp);
+  }
```
[the remark this answers is not in the document]
Comment (jflo) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadTest.java:288]: Previous test was covering that the exception had been caught, and wrapped into a Result. This is the case where it is not handled, and the exception is thrown upward of rememberBlock().

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinator.java [hunk 6]
```diff
@@ -99,7 +100,7 @@ public MergeCoordinator(
             // TODO: revisit default target gas limit
             .orElse(new AtomicLong(30000000L));
 
-    this.mergeBlockCreator =
+    this.mergeBlockCreatorFactory =
```
[the remark this answers is not in the document]
Comment (jflo) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinator.java:103]: Yes! however that requires a general refactoring of making MergeCoordinator much less aware of block production concerns (targetGasLimit, feeBeneficiary, extraData etc) so I will open a different PR to address that.

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinatorTest.java (hunk 1 above)
Comment (jflo) [reply consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinatorTest.java:301]: good catch! the intent was to make sure nothing was added to the badblock manager both before and after processing.  I've fixed this.

**Review by fab-10 — APPROVED** [review]

**Pull request merged**